File: cli-os/src/gateway/adapters/openaiCompat.js
// OpenAI-compatible passthrough adapter. Covers OpenAI itself and every provider that exposes
// an OpenAI-shaped /chat/completions (GLM/Zhipu, DeepSeek, Gemini's compat layer, Groq,
// Mistral, OpenRouter, local Ollama/vLLM). The request is already the canonical schema, so the
// work is: force streaming usage on, forward, and normalize usage out. Provider chunks are
// forwarded verbatim.
export const kind = 'openai-compat';

export function buildRequest({ model, openaiReq, stream }) {
  const body = { ...openaiReq, model };
  delete body.l00prite; // strip any gateway-only hints
  if (stream) {
    body.stream = true;
    body.stream_options = { ...(body.stream_options || {}), include_usage: true };
  } else {
    delete body.stream;
  }
  return body;
}

export const url = (baseUrl) => `${baseUrl.replace(/\/$/, '')}/chat/completions`;
export const headers = (apiKey) => ({ 'content-type': 'application/json', authorization: `Bearer ${apiKey}` });

function normUsage(u = {}) {
  return {
    prompt_tokens: u.prompt_tokens ?? 0,
    completion_tokens: u.completion_tokens ?? 0,
    cache_read_tokens: u.prompt_tokens_details?.cached_tokens ?? 0,
    cache_write_tokens: 0,
  };
}

export function parseFull(json, model) {
  // Already OpenAI-shaped; pass through (ensure model label present) and lift usage.
  if (!json.model) json.model = model;
  return { openaiResponse: json, usage: normUsage(json.usage) };
}

export function newStreamState(model) { return { model, usage: null }; }

export function onEvent(st, ev) {
  if (ev.data === '[DONE]') return { deltas: [], done: true, usage: st.usage || undefined };
  let chunk;
  try { chunk = JSON.parse(ev.data); } catch { return { deltas: [] }; }
  if (chunk.usage) st.usage = normUsage(chunk.usage);
  // A usage-only final chunk (choices=[]) carries no delta to forward downstream.
  const hasChoice = Array.isArray(chunk.choices) && chunk.choices.length;
  return { deltas: hasChoice ? [chunk] : [], usage: chunk.usage ? normUsage(chunk.usage) : undefined };
}

File: cli-os/internal/gateway/adapters/openaicompat.go
// OpenAI-compatible passthrough adapter. Covers OpenAI itself and every provider exposing an
// OpenAI-shaped /chat/completions (GLM/Zhipu, DeepSeek, Gemini compat, Groq, Mistral, OpenRouter,
// local Ollama/vLLM). The request is already canonical, so the work is: force streaming usage on,
// forward, and normalize usage out. Ported from openaiCompat.js.
package adapters

import (
	"encoding/json"
	"strings"

	"github.com/jackofall1232/l00prite/cli-os/internal/oai"
)

type openaiCompatAdapter struct{}

func (openaiCompatAdapter) Kind() string { return "openai-compat" }
func (openaiCompatAdapter) Direct() bool { return false }

func (openaiCompatAdapter) BuildRequest(model string, req map[string]any, stream bool) map[string]any {
	body := copyMap(req)
	body["model"] = model
	delete(body, "l00prite") // strip any gateway-only hints
	if stream {
		body["stream"] = true
		so := map[string]any{}
		if existing := asMap(body["stream_options"]); existing != nil {
			so = copyMap(existing)
		}
		so["include_usage"] = true
		body["stream_options"] = so
	} else {
		// stream_options is only valid alongside stream:true; leaving it on a non-stream request makes
		// strict providers (OpenAI) 400. Drop both.
		delete(body, "stream")
		delete(body, "stream_options")
	}
	return body
}

func (openaiCompatAdapter) URL(baseURL string) string {
	return strings.TrimSuffix(baseURL, "/") + "/chat/completions"
}

func (openaiCompatAdapter) Headers(apiKey string) map[string]string {
	return map[string]string{"content-type": "application/json", "authorization": "Bearer " + apiKey}
}

func normUsage(u map[string]any) oai.Usage {
	usage := oai.Usage{
		PromptTokens:     numToInt(u["prompt_tokens"]),
		CompletionTokens: numToInt(u["completion_tokens"]),
	}
	if details := asMap(u["prompt_tokens_details"]); details != nil {
		usage.CacheReadTokens = numToInt(details["cached_tokens"])
	}
	return usage
}

func (openaiCompatAdapter) ParseFull(resp map[string]any, model string) FullResult {
	if _, ok := resp["model"]; !ok || asStr(resp["model"]) == "" {
		resp["model"] = model
	}
	return FullResult{Response: resp, Usage: normUsage(asMap(resp["usage"]))}
}

func (openaiCompatAdapter) NewStream(model string, forwardUsage bool) StreamTranslator {
	return &openaiCompatStream{model: model, forwardUsage: forwardUsage}
}

type openaiCompatStream struct {
	model        string
	forwardUsage bool
	usage        *oai.Usage
}

func (s *openaiCompatStream) OnEvent(ev SSEEvent) StreamOut {
	if ev.Data == "[DONE]" {
		return StreamOut{Done: true, Usage: s.usage}
	}
	var chunk map[string]any
	if err := json.Unmarshal([]byte(ev.Data), &chunk); err != nil {
		return StreamOut{}
	}
	var evtUsage *oai.Usage
	if u := asMap(chunk["usage"]); u != nil {
		nu := normUsage(u)
		s.usage = &nu
		evtUsage = &nu
	}
	hasChoice := len(asArr(chunk["choices"])) > 0
	forward := hasChoice || (s.forwardUsage && evtUsage != nil)
	var deltas []map[string]any
	if forward {
		deltas = []map[string]any{chunk}
	}
	return StreamOut{Deltas: deltas, Usage: evtUsage}
}

// Usage returns the last usage seen (fallback when a stream ends before its terminal usage event).
func (s *openaiCompatStream) Usage() oai.Usage {
	if s.usage != nil {
		return *s.usage
	}
	return oai.Usage{}
}
